services/src/handlers/tasks.rs

```diff
@@ -124,29 +178,159 @@ mod tests {
     };
     use actix_http::header;
     use actix_web_httpauth::headers::authorization::Bearer;
-    use futures::channel::oneshot;
+    use futures::{channel::oneshot, lock::Mutex};
     use geoengine_datatypes::{error::ErrorSource, util::test::TestDefault};
+    use std::{pin::Pin, sync::Arc};
 
     struct NopTask {
-        complete_rx: oneshot::Receiver<()>,
+        complete_rx: Arc<Mutex<oneshot::Receiver<()>>>,
+        unique_id: Option<String>,
+    }
+
+    impl NopTask {
+        pub fn new_with_sender() -> (Self, oneshot::Sender<()>) {
+            let (complete_tx, complete_rx) = oneshot::channel();
+
+            let this = Self {
+                complete_rx: Arc::new(Mutex::new(complete_rx)),
+                unique_id: None,
+            };
+
+            (this, complete_tx)
+        }
+
+        pub fn new_with_sender_and_unique_id(unique_id: String) -> (Self, oneshot::Sender<()>) {
+            let (complete_tx, complete_rx) = oneshot::channel();
+
+            let this = Self {
+                complete_rx: Arc::new(Mutex::new(complete_rx)),
+                unique_id: Some(unique_id),
+            };
+
+            (this, complete_tx)
+        }
     }
 
     #[async_trait::async_trait]
     impl<C: TaskContext + 'static> Task<C> for NopTask {
-        async fn run(
-            self: Box<Self>,
-            _ctx: C,
-        ) -> Result<Box<dyn TaskStatusInfo>, Box<dyn ErrorSource>> {
-            self.complete_rx.await.unwrap();
+        async fn run(&self, _ctx: C) -> Result<Box<dyn TaskStatusInfo>, Box<dyn ErrorSource>> {
+            let mut complete_rx_lock = self.complete_rx.lock().await;
+            let pinned_receiver: Pin<&mut oneshot::Receiver<()>> = Pin::new(&mut complete_rx_lock);
+
+            pinned_receiver.await.unwrap();
+
             Ok("completed".to_string().boxed())
         }
 
+        async fn cleanup_on_error(&self, _ctx: C) -> Result<(), Box<dyn ErrorSource>> {
+            let mut complete_rx_lock = self.complete_rx.lock().await;
+            let pinned_receiver: Pin<&mut oneshot::Receiver<()>> = Pin::new(&mut complete_rx_lock);
+
+            pinned_receiver.await.unwrap();
+
+            Ok(())
+        }
+
         fn task_type(&self) -> &'static str {
             "nopTask"
         }
 
         fn task_unique_id(&self) -> Option<String> {
-            Some("highlander task".to_string())
+            self.unique_id.clone()
+        }
+    }
+
+    struct TaskWithChildren<T: TaskManager<C>, C: TaskContext + 'static> {
```

Comment:
would `TaskTree` be a better name?

Reply:
[43fe2ec](https://github.com/geo-engine/geoengine/pull/586/commits/43fe2ecd5494e62ee148cba3489ace27793a512d)